### DISCUSSION TITLE
Regenerate CSS & Data: Automatic Regeneration

### DISCUSSION THREAD

**Discussion by ghost** [elementor/elementor#28192] [Feature Request]
### Prerequisites

- [X] I have searched for similar feature requests in both open and closed discussions and cannot find an existing request.
- [X] I have verified that the feature is still missing from the latest stable versions of Elementor or Elementor Pro.

### What problem is your feature request going to solve? Please describe..

As an agency, updating Elementor can be extremely painful. After every update, I must log into several websites to manually regenerate the Elementor CSS and Data to fix design-related issues. 

### Describe the solution you'd like.

After every Elementor update, automatically regenerate the Elementor CSS & Data.

There's no harm in doing this for us. The alternative is far worse.

### Describe alternatives you've considered.

Not using Elementor.

### Agreement

- [X] I agree that my request may be closed without action if it doesn't meet all the requirements.

**Comment by jazir555** [comment]
Shouldn't this be the easiest functionality for Elementor to add ever? The functionality to regenerate the CSS already exists, this is essentially just automating it. Can the Elementor devs seriously not bang this out in like 2 hours? @nicholaszein

**Reply by nicholaszein** [reply]
@jazir555 thanks for asking this question. You'd be surprised to know that this already happens! 

I can explain how and why, but it's a really long read. I would be happy to write it though. 

If you're asking yourself why I haven't closed this discussion then, the simple answer is: because when it was open I didn't know this either, and had to really dive deep into the code to understand it myself before closing it without certainty.

Please let me know if you're interested in my explanation.

**Reply by jazir555** [reply]
Sure I'd like that. It sounds like you're saying that manually regenerating it isn't necessary since it's already done automatically if I'm interpreting your response correctly? If that is in fact the case, why are there numerous open issues where the posters have to explicitly manually regenerate the CSS to fix broken styles?

I can see only a couple potential possibilities. 

One is that the functionality for automatically regenerating it is in fact implemented, but it is bugged and there are edge cases where it isn't regenerating it, which is the source of these bug reports.

Two is that manually regenerating the CSS forces a cache clear and thus is really a placebo, and the underlying issue is actually cache purging.

Either way, something is bugged and needs to be fixed. Either the edge cases need to be identified and resolved, or more robust cache clearing mechanisms need to be added to invalidate stale content and display the newest version.

If in fact it is a cache purging issue, I would recommend looking into "The Cache Purger" to fully implement robust cache purging to ensure all caches are being purged simultaneously:

https://wordpress.org/plugins/the-cache-purger/

**Reply by nicholaszein** [reply]
Yep. It's a cache problem.

**Reply by ghost** [reply]
That plugin has 300ish active installs. Elementor powers how many millions of websites.?  The logic doesn’t make sense that now millions of websites need this plugin for Elementor-based design to NOT break. 

That cannot be the answer here, as out of those millions of sites, I guarantee that a ton of them have multiple levels of caching going on.

So, if Elementor knows this to be fact and they are not communicating it broadly, that’s yet another issue. 

A one-liner saying “yep, it’s a caching issue” is irresponsible.

**Reply by nicholaszein** [reply]
@wpturnedup

If you can wait for my explanation before making any assumptions, maybe we can find a solution to your issue that will solve it once and for all.

There is one, but without explaining it, I'm afraid you'll end up taking the wrong conclusions too. I hope not...

You can try the following:

In your case, this issue can be solved quickly by switching the **CSS Print Method** to **Internal Embedding** in the Elementor > Settings > Performance page. This is how Bricks loads theirs styles by default, and prevents caching plugins from caching pages without their styles. Since you opted to use cache on your sites, this would be the best approach.

But there is an explanation why this is important. It's not an arbitrary decision. It has to do with how cache works, and how the Elementor CSS engine works.

**Reply by nicholaszein** [reply]
@jazir555, I'm writing something up, and should be able to share it later today, or most likely tomorrow morning. It's actually something simple, but to understand how simple it is, I need to explain why it was designed this way. It has a lot to do with how WordPress and cache work.

**Reply by jazir555** [reply]
@nicholaszein 

Great, I would appreciate that. I think the issue is coming from out of sync caches since I have this issue all of the time, and in my case, the culprit is NGINX's cache. Since my whole jig is optimization I've tried just about every variation of optimization stacks and tried almost everything in the book, and in almost all cases for me, it's out of sync caches.

I would definitely recommend inquiring in the other threads if other people are using multi-layered caching, because I would bet money that's the source of the issues if this isn't related to the regenerate css function.

Also, my site is set to internal embedding and I still have the same problem. 

@wpturnedup 

The reason I'm suggesting using The Cache Purger is to verify the problem. If you could just install it temporarily to confirm this is the issue, it will help the Elementor team implement a fix by narrowing down the issue.

It's simply a diagnostic step. If you don't want to even test it out of principle, I'm not exactly sure what you're looking for here, or how you expect to receive help. If you don't want to provide them with the information they need there's not much they can do to help you.

If you can verify it's a caching issue, that indicates that Elementor is not clearing all caches simultaneously upon publishing a page, that's critical information that will be used to design a fix.

**Reply by nicholaszein** [reply]
@jazir555 I'm actually considering making this a blog post in https://developers.elementor.com/blog. But it's just an idea for now. There is more to it, like for example cron jobs, and the fact that Elementor does have hooks to a few cache plugins. WP Rocket for example has tailored their plugin to work with Elementor, granted you follow their instructions. As you said yourself, it's a multi-layered, multi-combinations type of problem, and unfortunately not all cache solutions (mainly server-side) can play nice with Elementor. Having the CSS embedded helps a lot, but as you pointed out, it may not be enough for some cases (object cache for example can affect the output). In a perfect world everything would simply work together, but with so many different levels and solutions, its virtually impossible to predict every combination. One of the worst "offenders" in this vast helm of performance optimization solutions is, for example LiteSpeed. It works perfectly fine for static sites that never change nor are regularly updated. But that's not the reality of a good portion of WordPress websites. WordPress is dynamic by nature.

I confess, I'm not a fast writer, but I'll do my best to be as clear as possible, and try to simplify thigs as much as I can with my writing.

**Reply by jazir555** [reply]
@nicholaszein 

Thanks, I think making a blog post is a good idea. I can't recommend looking into The Cache Purger enough since this is basically its entire purpose. There are tons of caching systems, and its exclusive focus is clearing every known cache mechanism. I think it would provide a good framework for improving Elementor's cache clearing functions, and the dev is responsive and I'm sure would be happy to collaborate with you guys.

**Reply by nicholaszein** [reply]
Hey @jazir555!

It seems I completely underestimated the time it would take to write this, and I still haven't finished the text. I think I need to take some real time on it over the weekend. Unfortunately, @wpturnedup has either blocked me or deleted his account, but I hope he can read this later.

**Reply by jazir555** [reply]
@nicholaszein 

It looks like he deleted his account, it displays as "ghost" even in incognito.

>It seems I completely underestimated the time it would take to write this, and I still haven't finished the text. I think I need to take some real time on it over the weekend

No worries, I've had the same where I've significantly underestimated the time required to do a thorough write up of something. I'll look forward to the post hopefully sometime this week or next week :)

**Reply by nicholaszein** [reply]
Thank you @jazir555! 🙏 

I'm almost finished with the article, and after that we'll have a period of review from my colleagues. I'm hoping to publish it as an article soon, but I don't have a date yet. If it takes too long I'll share it here. I'll keep you posted! Thank you for being so kind and patient! ❤️

**Reply by jazir555** [reply]
Sure no problem! Looking forward to it!

**Comment by ghost** [comment]
It’s not a bug. It’s not going to be a feature. So, file another complaint with tech support.?  I’m on it!

Even if this is not Elementor’s “fault,” it needs to be a discussion about why it’s happening, so Elementor’s customers can get to the bottom of it.

I’ll continue to say it. Elementor’s GitHub management needs to be addressed.  It’s seriously flawed.

**Reply by jazir555** [reply]
I would assume if this isn't in fact Elementor being bugged and is regenerating the CSS automatically, this is indicative of a cache purging issue. I would recommend installing The Cache Purger and seeing if the issue is resolved. If it is, then this is assuredly out of sync caches because one cache is not being purged correctly.

If that is the case, css regeneration is not the issue.

**Reply by ghost** [reply]
Purging the cache has never worked for me, when it comes to Elementor-based content/design being broken after and Elementor update.

Only the Elementor regeneration process fixes the issues.

**Reply by jazir555** [reply]
If you're using multi-layered caching that would be why. Object caching, Page caching, CDN cache, Server cache, Varnish, etc could be out of sync. I have to manually purge my nginx cache to fix the issue.

Try The Cache Purger as it will ensure all caches are purged simultaneously.

**Reply by ghost** [reply]
For fun, I just went to see what Bricks is doing in this area. They made this a feature within their builder over a year ago.

**Reply by jazir555** [reply]
Still, can you try The Cache Purger to see if it resolved the issue? I'd bet this is it, I've had the same problem with broken styles, which in my case was cache purging related.

**Reply by nicholaszein** [reply]
@wpturnedup I completely understand your frustration. All I ask is that you don't make conclusions just yet.

You seem to be upset specifically with me, and I understand that. If you ever felt mistreated in any way by me, I sincerely apologize. 

Like you, I'm here because I care, but it takes time to make things right, and I cannot yet give you all the attention and care you expect and deserve. I'm not, in any way looking to antagonize and dismiss your submissions and feedback. They are actually of extreme value for mee and for the rest of the team. Like in any big company, we need to organize things, and separate them, so we can address them in an orderly fashion, otherwise we would never be able to address anything. This is why there are procedures, and places specific for each type of communication, and I am required to enforce them for the sake of everyone. 

In any way, we do recognize GitHub needs more organization, and we've been working hard for quite some time to achieve this goal. Our commitment is to address each and every message, so everyone can feel heard. Unfortunately, it's not an easy task that can be solved quickly. We're just not there yet. There are quick ways to solve this, but they are not beneficial for users like yourself, and that's not our goal.
